### PR TITLE
Change representation for fingerprints

### PR DESCRIPTION
**Pull request by GambolingPangolin**
This changes how we represent fingerprints to `newtype Fingerprint = Fingerprint { unFingerprint :: Word32 }`.  This has a few benefits.  First, `Fingerprint` has `Serialize` etc. instances and downstream structures use these instances, so there is no risk of the encoding drifting from one place to another.  (The binary encodings for fingerprints in the extended keys section disagreed with that from the psbt section.)   Second, I added IsString/Read/Show instances that treat `Fingerprint` as having an 8-digit hex string representation instead of a decimal number.

### CODE REVIEW

File: src/Haskoin/Transaction/Partial.hs [hunk 1]
```diff
@@ -679,15 +679,15 @@ instance Serialize PSBTHDPath where
...
+                ((,) <$> S.getWord32be <*> getKeyIndexList numIndices)
       where
         getKeyIndexList n = replicateM n S.getWord32le
 
     put (PSBTHDPath (fp, kis)) = do
         putVarInt (B.length bs)
         S.putByteString bs
       where
-        bs = S.runPut $ S.putWord32le fp >> mapM_ S.putWord32le kis
+        bs = S.runPut $ S.putWord32be fp >> mapM_ S.putWord32le kis
```
Comment (wraithm) [inline src/Haskoin/Transaction/Partial.hs]: What about the other word32le there?

Is there mixed byteorder?

Also, I'm surprised the test suite doesn't catch this!!

Would you mind checking if the psbt test suite is missing vectors?

Comment (GambolingPangolin) [reply src/Haskoin/Transaction/Partial.hs]: We might also need to change that.  `DerivPathI` uses big-endian byte order for key indices.  I'll add test vectors as needed and make sure that what we do conforms with the BIP 174 spec.  The BIP has a lot of test vectors.

Comment (GambolingPangolin) [reply src/Haskoin/Transaction/Partial.hs]: The situation is messy:
- BIP 32 specifies big-endian for key indices
- BIP 174 specifies little-endian for HD path components

Also, I added a test case covering HD paths.

**Comment by GambolingPangolin** [comment]
Another possibility is to put `Fingerprint` behind an abstract `newtype`, using `Word32` as an opaque implementation detail.  Then if downstream (de)serializers used its serialization instances there would be no danger of divergence.  Also, we could make the show instance be the hex representation.

File: test/Haskoin/Transaction/PartialSpec.hs [hunk 2]
```diff
@@ -137,6 +203,7 @@ vec6Test = do
         (Key 0x0f (fromJust $ decodeHex "010203040506070809"))
         (fromJust $ decodeHex "0102030405060708090a0b0c0d0e0f")
 
+
```
Comment (wraithm) [inline test/Haskoin/Transaction/PartialSpec.hs]: nit: unnecessary whitespace

File: test/Haskoin/Transaction/PartialSpec.hs [hunk 3]
```diff
@@ -121,6 +119,74 @@ vec5Test = do
...
+                   ]
+        , outputs = [ Output { outputRedeemScript = Nothing
+                             , outputWitnessScript = Nothing
+                             , outputHDKeypaths = mempty
+                             , outputUnknown = mempty
+                             }
+                    ]
+        }
+    hardIndex = (+ 2^31)
+
```
Comment (wraithm) [inline test/Haskoin/Transaction/PartialSpec.hs:190]: nit: unnecessary whitespace

**Comment by GambolingPangolin** [comment]
I went ahead and changed `Fingerprint` to a `newtype`.

File: haskoin-core.cabal [hunk 4]
```diff
@@ -86,7 +86,7 @@ library
       Haskoin.Util.Arbitrary.Transaction
       Haskoin.Util.Arbitrary.Util
   other-modules:
-      Paths_haskoin_core
```
Comment (wraithm) [inline haskoin-core.cabal:89]: Why remove this?

Comment (wraithm) [reply haskoin-core.cabal:89]: Oh, it's automatic! hmm

File: src/Haskoin/Keys/Extended/Internal.hs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+instance IsString Fingerprint where
+    fromString = fromRight decodeError
+        . S.decode
+        . fromMaybe hexError
+        . decodeHex
+        . Text.pack
+     where
+       decodeError  = error "Fingerprint literal: Unable to decode"
+       hexError = error "Fingerprint literal: Invalid hex"
+
```
Comment (wraithm) [inline src/Haskoin/Keys/Extended/Internal.hs:49]: nit: 2 newlines

File: haskoin-core.cabal (hunk 4 above)
Comment (GambolingPangolin) [reply haskoin-core.cabal:89]: This must have been a side effect of _adding_ an `other-modules` field to `package.yaml`.  Do we use the paths module for anything?  The project builds without it.  Maybe we should leave it off.

Comment (wraithm) [reply haskoin-core.cabal:89]: Yeah, let's maybe leave it off. The paths modules cause all kinds of problems with test rebuilding and stuff... Would you mind actually explicitly ignoring it (like we do in our repo)?

Comment (GambolingPangolin) [reply haskoin-core.cabal:89]: Sure!

File: CHANGELOG.md [hunk 6]
```diff
@@ -4,6 +4,11 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
+## 0.21.0
+### Changed
+- Use a newtype for Fingerprint, which uses an 8 digit hex string for IsString,
+  Read, & Show.  This fixes inconsistent (de)serialization across the package.
```
Comment (wraithm) [inline CHANGELOG.md:10]: It might be good to mention the Paths_haskoin_core thing.

@jprupp, what do you think about this? I don't imagine that people are depending on this module, but you can tell me otherwise.

Comment (wraithm) [reply CHANGELOG.md:10]: We can also take the conservative route and still expose it.

Comment (GambolingPangolin) [reply CHANGELOG.md:10]: `Paths_haskoin_core` was not part of the public API.  It was under `other-modules` and now it's unlisted altogether.

Comment (wraithm) [reply CHANGELOG.md:10]: Ah, okay! Nvm then. Resolving.

**Review by wraithm — APPROVED** [review]

**Comment by wraithm** [comment]
@jprupp, Do you want to take a look at this? I confirmed that this is now ready to go.

**Comment by jprupp** [comment]
Thanks @wraithm. Looks good to me.
